bybit: enable multiconnection handling across websocket endpoints

shazbert opened the pull request:
# PR Description

* Expand websocket handling to concurrently allow processing for CoinMarginedFutures, USDTMarginedFutures, USDCMarginedFutures, Options. 
* Implement dedicated connection for private data.

NOTE: This is a nice size PR for just functionality expansion, I will build outbound request functions for trading on top of this in a future PR. Here's the dependency: #1603


Fixes # 

* 🐛 ~DOGEUSDT wasn't being processed correctly so used `MatchSymbolWithAvailablePairs` because there was no delimiter.~ This was merged upstream.

## Type of change

Please delete options that are not relevant and add an `x` in `[]` as item is complete.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How has this been tested

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration and
also consider improving test coverage whilst working on a certain feature or package.

- [ ] go test ./... -race
- [ ] golangci-lint run
- [ ] Test X

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally and on Github Actions with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

gloriousCode commented on exchanges/bybit/bybit_websocket.go:
```diff
@@ -792,3 +740,162 @@ const subTplText = `
+	default:
+		return currency.EMPTYPAIR, 0, fmt.Errorf("category '%s' not supported for incoming symbol '%s'", category, symbol)
+	}
+	var err error
+	for _, a := range assets {
+		var cp currency.Pair
+		if cp, err = by.MatchSymbolWithAvailablePairs(symbol, a, hasPotentialDelimiter(a)); err == nil {
+			return cp, a, nil
+		}
+	}
+	return currency.EMPTYPAIR, 0, err
```
```go
	for _, a := range assets {
		cp, err := by.MatchSymbolWithAvailablePairs(symbol, a, hasPotentialDelimiter(a))
		if err != nil {
			return currency.EMPTYPAIR, 0, err
		}
		return cp, a, nil
	}
	return currency.EMPTYPAIR, 0, nil
```
I think its an improvement to readability to not try funny business with err handling and such

gloriousCode commented on exchanges/bybit/bybit_websocket.go:
```diff
@@ -792,3 +740,162 @@ const subTplText = `
+	for _, channel := range []string{chanPositions, chanExecution, chanOrder, chanWallet} {
+		subscriptions = append(subscriptions, &subscription.Subscription{
+			Channel: channel,
+			Pairs:   currency.Pairs{currency.EMPTYPAIR}, // This is a placeholder, the actual pair is not required for these channels
```
Is pairs required to be filled? Have you tried not sending pairs?

gloriousCode reviewed (changes requested):
Nice work shazbert, I'm quite happy with a lot of the work

shazbert commented on exchanges/bybit/bybit_websocket.go:
```diff
@@ -313,14 +240,51 @@ func (by *Bybit) wsHandleData(assetType asset.Item, respRaw []byte) error {
+		return by.wsProcessExecution(&result)
 	case chanOrder:
-		return by.wsProcessOrder(asset.Spot, &result)
+		// Below provides a way of matching an order change to a websocket request. There is no batch support for this
+		// so the first element will be used to match the order ID.
+		if id, err := jsonparser.GetString(respRaw, "data", "[0]", "orderId"); err == nil {
```
Nice catch this was actually updated to that in the request PR built on this. 🚀 I will roll that update here.

shazbert commented on exchanges/bybit/bybit_websocket.go:
```diff
@@ -792,3 +740,162 @@ const subTplText = `
+	for _, payload := range payloads {
+		if a == asset.Options {
+			// The options connection does not send the subscription request id back with the subscription notification payload
+			// therefore the code doesn't wait for the response to check whether the subscription is successful or not.
```
Going to be RM'd anyway

gbjk commented on exchanges/bybit/bybit_test.go:
```diff
@@ -3788,46 +3780,47 @@ func TestGenerateSubscriptions(t *testing.T) {
+	require.NoError(t, b.authSubscribe(t.Context(), &DummyConnection{}, subscription.List{}))
+
+	authsubs, err := b.generateAuthSubscriptions()
+	require.NoError(t, err)
+	require.Empty(t, authsubs)
```
🐒 🌴 
```suggestion
	require.NoError(t, err, "generateAuthSubscriptions must not error")
	require.Empty(t, authsubs, "generateAuthSubscriptions must not return subs")
```

gbjk commented on exchanges/bybit/bybit_test.go:
```diff
@@ -3788,46 +3780,47 @@ func TestGenerateSubscriptions(t *testing.T) {
-	err = b.Subscribe(subs)
-	assert.ErrorContains(t, err, "Mock Resp Error", "Subscribe should error containing the returned RetMsg")
+	authsubs, err = b.generateAuthSubscriptions()
+	require.NoError(t, err)
+	require.NotEmpty(t, authsubs)
```
And throughout.
```suggestion
	require.NoError(t, err, "generateAuthSubscriptions must not error")
	require.Empty(t, authsubs, "generateAuthSubscriptions must not return subs")
```

gbjk commented on exchanges/bybit/bybit_test.go:
```diff
@@ -3893,3 +3886,89 @@ func TestTransformSymbol(t *testing.T) {
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "unexpected end of JSON input")
+
+	wsOrderResponse = WebsocketResponse{
+		Data: []byte(`[{"category":"linear","symbol":"BTCUSDT","orderId":"c1956690-b731-4191-97c0-94b00422231b","orderLinkId":"","blockTradeId":"","side":"Sell","positionIdx":0,"orderStatus":"Filled","cancelType":"UNKNOWN","rejectReason":"EC_NoError","timeInForce":"IOC","isLeverage":"","price":"4.033","qty":"1.7","avgPrice":"4.24","leavesQty":"0","leavesValue":"0","cumExecQty":"1.7","cumExecValue":"7.2 ...
+	}
+
+	var o []order.Detail
+	var wg sync.WaitGroup
+	var ok bool
+	wg.Add(1)
+	go func() {
+		data := <-b.Websocket.DataHandler
+		o, ok = data.([]order.Detail)
+		wg.Done()
+	}()
+	require.NoError(t, b.wsProcessOrder(&wsOrderResponse))
+	wg.Wait()
+	require.True(t, ok)
+	require.Len(t, o, 1)
+	require.Equal(t, "c1956690-b731-4191-97c0-94b00422231b", o[0].OrderID)
+	require.Equal(t, "BTC_USDT", o[0].Pair.String())
+	require.Equal(t, order.Sell, o[0].Side)
+	require.Equal(t, order.Filled, o[0].Status)
+	require.Equal(t, 1.7, o[0].Amount)
+	require.Equal(t, 4.033, o[0].Price)
+	require.Equal(t, 4.24, o[0].AverageExecutedPrice)
+	require.Equal(t, 0., o[0].RemainingAmount)
+	require.Equal(t, asset.USDTMarginedFutures, o[0].AssetType)
```
🐒  Also here; Can this not use the FixtureToDataHandler pattern?

shazbert replied:
[removed and consolidated in](https://github.com/thrasher-corp/gocryptotrader/pull/1670/commits/49930d2190a74540225f1d89737ac66039fa6cc7)

gbjk reviewed (approved):
Good work 👍 

I'm generally unhappy with needing to do any of this, but I can see the constraints, and if we spend more time hacking out anything better right now we might as well spend that time on the already open branches.

I've downgraded everything to nitpicks from change requests because this is high priority, and you can fix them later. None of them introduce bugs or regressions.

shazbert commented on exchange/websocket/connection.go:
```diff
@@ -111,23 +111,23 @@ type Response struct {
+	ResponseMaxLimit     time.Duration
+	Traffic              chan struct{}
+	readMessageErrors    chan error
+	requestIDGenerator   func() int64
```
Good suggestion, I think I might make a separate PR for this, because this will impact GenerateMessageID signature as well, I would probably opt for `defaultGenerateMessageID` to be v7 for searchability and `ClientOrderID` field setting. There is a tiny overhead associated with v7 but it's minute, so keeping it as a baseline would be perfect. The current default is just using `rand.Int` and has the potential to have collisions anyway. 

For GateIO/bybit the counter is used mainly for request response matching and the counters don't seem to be tracked or are reset on each new connection and they don't come back in fills/order history. So if possible would like to keep them as is and then refresh them if needed in that separate PR?

Thoughts? @thrasher- @gbjk

gbjk replied:
Agreed to slide this to later.

shazbert commented on exchanges/bybit/bybit_test.go:
```diff
@@ -3211,21 +3188,36 @@ var pushDataMap = map[string]string{
+	"Private Order":    `{ "id": "5923240c6880ab-c59f-420b-9adb-3639adc9dd90", "topic": "order", "creationTime": 1672364262474, "data": [ { "symbol": "%s", "orderId": "5cf98598-39a7-459e-97bf-76ca765ee020", "side": "Sell", "orderType": "Market", "cancelType": "UNKNOWN", "price": "72.5", "qty": "1", "orderIv": "", "timeInForce": "IOC", "orderStatus": "Filled", "orderLinkId": "", "lastPriceOnCreated": ...
+	"Private Wallet":   `{ "id": "5923242c464be9-25ca-483d-a743-c60101fc656f", "topic": "wallet", "creationTime": 1672364262482, "data": [ { "accountIMRate": "0.016", "accountMMRate": "0.003", "totalEquity": "12837.78330098", "totalWalletBalance": "12840.4045924", "totalMarginBalance": "12837.78330188", "totalAvailableBalance": "12632.05767702", "totalPerpUPL": "-2.62129051", "totalInitialMargin": " ...
+	"Private Greek":    `{ "id": "592324fa945a30-2603-49a5-b865-21668c29f2a6", "topic": "greeks", "creationTime": 1672364262482, "data": [ { "baseCoin": "ETH", "totalDelta": "0.06999986", "totalGamma": "-0.00000001", "totalVega": "-0.00000024", "totalTheta": "0.00001314" } ] }`,
+	"Execution":        `{"id": "592324803b2785-26fa-4214-9963-bdd4727f07be", "topic": "execution", "creationTime": 1672364174455, "data": [ { "category": "linear", "symbol": "XRPUSDT", "execFee": "0.005061", "execId": "7e2ae69c-4edf-5800-a352-893d52b446aa", "execPrice": "0.3374", "execQty": "25", "execType": "Trade", "execValue": "8.435", "isMaker": false, "feeRate": "0.0006", "tradeIv": "", "markI ...
```
https://github.com/thrasher-corp/gocryptotrader/pull/1670/commits/49930d2190a74540225f1d89737ac66039fa6cc7

shazbert commented:
> I'm generally unhappy with needing to do any of this

![image](https://github.com/user-attachments/assets/fd4eacf7-e971-4145-aa2f-bc39cb0424aa)

gbjk commented on exchanges/gateio/gateio.go:
```diff
@@ -187,8 +187,9 @@ func timeInForceFromString(tif string) (order.TimeInForce, error) {
-	Counter common.Counter // Must be first	due to alignment requirements
 	exchange.Base
+
+	messageIDSeq  common.Counter
```
Agreed to slide this to later following [this reply](https://github.com/thrasher-corp/gocryptotrader/pull/1670#discussion_r2149088220)

gbjk commented on exchanges/bybit/bybit_test.go:
```diff
@@ -3060,6 +3036,172 @@ func TestPushData(t *testing.T) {
+	b.API.AuthenticatedWebsocketSupport = true
+	b.SetCredentials("test", "test", "", "", "", "")
+	testexch.FixtureToDataHandler(t, "testdata/wsAuth.json", func(ctx context.Context, r []byte) error {
+		if bytes.Contains(r, []byte("%s")) {
+			r = fmt.Appendf(nil, string(r), optionsTradablePair.String())
+		}
```
🥃 🗻  This feels like a generic problem.
Wondering if we can text/template the content easily, and introduce a middleware:
```
FixtureToDataHandler(t, "foo.json", testexch.WrapFixture(b.wsHandleAuthenticatedData), testexch.FixtureVars{
    "optionsPair": optionsTradablePair,
}
```
and then in the fixture json:
```
{{ .vars.optionsPair.String }}
```

shazbert replied:
Yeah that would be cool to do, can do this in another PR.

gbjk commented on exchanges/bybit/bybit_test.go:
```diff
@@ -3060,6 +3036,172 @@ func TestPushData(t *testing.T) {
+		return b.wsHandleAuthenticatedData(ctx, r)
+	})
+	close(b.Websocket.DataHandler)
+	require.Len(t, b.Websocket.DataHandler, 6, "Should see correct number of private messages")
```
```suggestion
	require.Len(t, b.Websocket.DataHandler, 6, "Should see correct number of messages")
```

gbjk commented on exchanges/bybit/bybit_test.go:
```diff
@@ -3060,6 +3036,172 @@ func TestPushData(t *testing.T) {
+			require.Len(t, v, 1, "must see 1 position")
+			assert.Zero(t, v[0].PositionIdx, "PositionIdx should be 0")
+			assert.Zero(t, v[0].TradeMode, "TradeMode should be 0")
+			assert.Equal(t, int64(41), v[0].RiskID, "RiskID should match")
```
🦍  The wording I've used so far is "X should be correct".
"match" feels overloaded, because of times we actually mean that some function or method called match returned true, etc.
Don't have to action, I'm just being pedantic/consistent.

gbjk commented on exchanges/bybit/bybit_test.go:
```diff
@@ -3060,6 +3036,172 @@ func TestPushData(t *testing.T) {
+			assert.Equal(t, 0.3374, v[0].Price, "price should match")
+			assert.Equal(t, 25.0, v[0].Amount, "amount should match")
+		default:
+			t.Errorf("Unexpected data type received: %T", v)
```
```suggestion
			t.Errorf("Unexpected data received: %v", v)
```

gbjk reviewed (approved):
All good.
Only wildlife to see. 🐒 🦍

cranktakular reviewed (changes requested):
Got some changes I want.

On top of those comments about specific places, some more general things:
* When running all exchange tests, I got an error in okx I didn't get on master (line 333, TestGetBlockTrade, "0 is not positive%!"). Find out whether that was caused by your changes, or if it's just intermittent.
* Potentially OOS, but consider writing detailed test functions to get more coverage out of GenerateInverseDefaultSubscriptions, InverseUnsubscribe, GenerateLinearDefaultSubscriptions, LinearUnsubscribe, GenerateOptionsDefaultSubscriptions, OptionUnsubscribe, Unsubscribe, and Setup.
* Potentially OOS, but I saw some room for more coverage in handleSubscriptions, handleSpotSubscription, wsHandleData, wsProcessWalletPushData, wsProcessOrder, wsProcessExecution, wsProcessLeverageTokenTicker, wsProcessPublicTicker, GetWalletBalance, and UpdateAccountInfo.
* I also saw some coverage shortfalls that I think are in scope. Within the functions wsHandleAuthenticatedData, wsProcessOrder, wsProcessPublicTicker, submitDirectSubscription, directSubscriptionPayload, getPairFromCategory, WsConnect, and WebsocketAuthenticatedConnection.

shazbert commented on exchanges/bybit/bybit_wrapper.go:
```diff
@@ -214,65 +209,137 @@ func (by *Bybit) SetDefaults() {
+		ResponseMaxLimit:      exch.WebsocketResponseMaxLimit,
+		RateLimit:             request.NewWeightedRateLimitByDuration(time.Microsecond),
+		Connector:             by.WsConnect,
+		GenerateSubscriptions: func() (subscription.List, error) { return by.generateSubscriptions() },
```
nice catch!

shazbert commented on exchanges/bybit/bybit_websocket.go:
```diff
@@ -793,3 +743,165 @@ const subTplText = `
+			authArg.associatedSubs = append(authArg.associatedSubs, s)
+		}
+
+		if len(arg.Arguments) >= 10 {
```
Can defer this to https://github.com/thrasher-corp/gocryptotrader/pull/1672 if there are any issues when trading on it. But yeah it should be removed entirely as soon as possible.

shazbert commented on exchanges/bybit/bybit_websocket.go:
```diff
@@ -793,3 +743,165 @@ const subTplText = `
+			if chanMap[s.Channel] {
+				continue
+			}
+			authArg.Arguments = append(authArg.Arguments, s.Channel)
```
Doesn't look like we handle DCP at all in `generateAuthSubscriptions`, added a TODO for implementation maybe for https://github.com/thrasher-corp/gocryptotrader/pull/1672 or when full template pass comes in.

shazbert replied:
What would you like me to do, that is in scope for this PR?

shazbert commented on exchanges/bybit/bybit_websocket.go:
```diff
@@ -176,60 +139,41 @@ func (by *Bybit) handleSubscriptions(operation string, subs subscription.List) (
-func (by *Bybit) handleSpotSubscription(ctx context.Context, operation string, channelsToSubscribe subscription.List) error {
-	payloads, err := by.handleSubscriptions(operation, channelsToSubscribe)
+func (by *Bybit) handleSpotSubscription(ctx context.Context, conn websocket.Connection, operation string, channelsToSubscribe subscription.List) error {
+	payloads, err := by.handleSubscriptions(conn, operation, channelsToSubscribe)
```
I was assuming it would be used across templates when that upgrade occurs. Suggest a better name I could implement.

cranktakular replied:
Ah, if it's pending that sort of thing then don't worry.

shazbert commented on exchanges/bybit/bybit_wrapper.go:
```diff
@@ -214,65 +209,137 @@ func (by *Bybit) SetDefaults() {
 		return err
 	}
 
-	wsRunningEndpoint, err := by.API.Endpoints.GetURL(exchange.WebsocketSpot)
```
Good catch will re-implement

shazbert commented on exchanges/bybit/bybit_websocket.go:
```diff
@@ -569,47 +531,35 @@ func (by *Bybit) wsProcessKline(assetType asset.Item, resp *WebsocketResponse, t
-	} else {
-		// ticker updates may be partial, so we need to update the current ticker
-		tick, err = ticker.GetTicker(by.Name, p, assetType)
+	tick := &ticker.Price{Pair: p, ExchangeName: by.Name, AssetType: assetType}
```
I might defer this, I think that @thrasher- has an update incoming for this on one of his PR's

shazbert commented:
@cranktakular most nits addressed thank you!

shazbert commented:
@cranktakular Still have more things to do but time constrained hold off on reviewing this just yet until I can get around to it.

shazbert commented:
- [ ]  websocket reconnection is broken, need to investigate. Something to do with ping handlers not being cleaned up. 🪦

shazbert commented:
> * [ ]   websocket reconnection is broken, need to investigate. Something to do with ping handlers not being cleaned up. 🪦

nevermind that was bad code from me. 
![image](https://github.com/user-attachments/assets/789c8dc6-57e9-44e3-bce6-ae43628bfe19)

shazbert commented:
@cranktakular 

> * When running all exchange tests, I got an error in okx I didn't get on master (line 333, TestGetBlockTrade, "0 is not positive%!"). Find out whether that was caused by your changes, or if it's just intermittent.

This PR doesn't impact that. 

> * Potentially OOS, but consider writing detailed test functions to get more coverage out of GenerateInverseDefaultSubscriptions, InverseUnsubscribe, GenerateLinearDefaultSubscriptions, LinearUnsubscribe, GenerateOptionsDefaultSubscriptions, OptionUnsubscribe, Unsubscribe, and Setup.

Added minor coverage improvements

> * Potentially OOS, but I saw some room for more coverage in handleSubscriptions, handleSpotSubscription, wsHandleData, wsProcessWalletPushData, wsProcessOrder, wsProcessExecution, wsProcessLeverageTokenTicker, wsProcessPublicTicker, GetWalletBalance, and UpdateAccountInfo.

Added minor coverage to this

> * I also saw some coverage shortfalls that I think are in scope. Within the functions wsHandleAuthenticatedData, wsProcessOrder, wsProcessPublicTicker, submitDirectSubscription, directSubscriptionPayload, getPairFromCategory, WsConnect, and WebsocketAuthenticatedConnection.

Added minor coverage for these, there are some which are going to be removed in template pass, `getPairFromCategory` will only error on nil currency manager so not going to hit that

samuael commented on exchanges/bybit/bybit_test.go:
```diff
@@ -3015,42 +3019,34 @@ func TestCancelBatchOrders(t *testing.T) {
+	require.NoError(t, err)
+	exp := errors.New("dial error")
+	err = e.WsConnect(t.Context(), &DummyConnection{dialError: exp})
+	require.ErrorIs(t, err, exp)
```
Could you use a mere string comparison instead?

samuael commented on exchanges/bybit/spot_websocket.go:
```diff
@@ -0,0 +1,50 @@
+			return err
+		}
+		var resp SubscriptionResponse
+		if err := json.Unmarshal(response, &resp); err != nil {
```
You can remove this immediate assignment since there is a big scoped err variable above.

shazbert replied:
prefer if short, won't resolve so people can yell at me.

samuael reviewed (approved):
Just two optional code suggestion spotted

cranktakular reviewed (changes requested):
Got some more changes I want.

In addition to those comments, I'm pretty sure these functions could get more test coverage: handleSpotSubscription, wsHandleAuthenticatedData, handleNoTopicWebsocketResponse, wsProcessOrder, getPairFromCategory, Setup.

And I think these functions have a slight chance of being able to get more test coverage: GenerateOptionsDefaultSubscriptions, GenerateLinearDefaultSubscriptions, GenerateInverseDefaultSubscriptions

shazbert commented on exchanges/bybit/bybit_websocket.go:
```diff
@@ -318,36 +187,68 @@ func (e *Exchange) wsHandleData(ctx context.Context, assetType asset.Item, respR
+		return e.handleNoTopicWebsocketResponse(&result, respRaw)
+	}
+	topicSplit := strings.Split(result.Topic, ".")
+	switch topicSplit[0] {
```
Will pass on this due to time constraints. You can open a PR against this merged state if it bugs you.

shazbert commented on exchanges/bybit/bybit_websocket.go:
```diff
@@ -318,36 +187,68 @@ func (e *Exchange) wsHandleData(ctx context.Context, assetType asset.Item, respR
+		// Below provides a way of matching an order change to a websocket request. There is no batch support for this
+		// so the first element will be used to match the order link ID.
+		if id, err := jsonparser.GetString(respRaw, "data", "[0]", "orderLinkId"); err == nil {
+			if e.Websocket.Match.IncomingWithData(id, respRaw) {
```
Will be RM'd for the appropriate PR #1672

shazbert commented on exchanges/bybit/inverse_websocket_test.go:
```diff
@@ -0,0 +1,58 @@
+
+func TestGenerateInverseDefaultSubscriptions(t *testing.T) {
+	t.Parallel()
+	e := new(Exchange) //nolint:govet // Intentional shadow
```
Yeah it will probably resolve itself when they update upstream. Until then.

shazbert commented on exchanges/bybit/bybit_test.go:
```diff
@@ -3623,48 +3791,43 @@ func TestGenerateSubscriptions(t *testing.T) {
-	require.NoError(t, err, "Subscribe must not error")
+
+	e := new(Exchange) //nolint:govet // Intentional shadow
+	require.NoError(t, testexch.Setup(e))
```
Yeah I generally do not include them because it has a line failure number. But you make a good point.

shazbert commented on exchanges/bybit/bybit_websocket.go:
```diff
@@ -63,94 +65,56 @@ var defaultSubscriptions = subscription.List{
 // WsConnect connects to a websocket feed
-func (by *Bybit) WsConnect() error {
-	ctx := context.TODO()
-	if !by.Websocket.IsEnabled() || !by.IsEnabled() || !by.IsAssetWebsocketSupported(asset.Spot) {
```
Yeah they won't generate subs, and having no subs means that no connection will be spawned and this function won't be called.

shazbert commented:
> Got some more changes I want.
> 
> In addition to those comments, I'm pretty sure these functions could get more test coverage: handleSpotSubscription, wsHandleAuthenticatedData, handleNoTopicWebsocketResponse, wsProcessOrder, getPairFromCategory, Setup.
> 
> And I think these functions have a slight chance of being able to get more test coverage: GenerateOptionsDefaultSubscriptions, GenerateLinearDefaultSubscriptions, GenerateInverseDefaultSubscriptions

Will pass on this due to time constraints. You can open a PR against this merged state if it bugs you. These generally have sufficient coverage and this need to be merged sooner rather than later.

cranktakular reviewed (approved):
Just got one suggested change left, other than that I can tACK this.

shazbert commented on exchanges/exchange.go:
```diff
@@ -1380,6 +1379,16 @@ func (u URL) String() string {
 		return restSwapURL
 	case WebsocketSpot:
 		return websocketSpotURL
+	case WebsocketCoinMargined:
```
You are welcome, but thank mr @cranktakular for kicking my butt on this. I originally just purged this completely.

shazbert commented on exchanges/bybit/bybit_websocket.go:
```diff
@@ -63,94 +64,58 @@ var defaultSubscriptions = subscription.List{
-	{Enabled: true, Asset: asset.Spot, Authenticated: true, Channel: subscription.MyOrdersChannel},
-	{Enabled: true, Asset: asset.Spot, Authenticated: true, Channel: subscription.MyWalletChannel},
-	{Enabled: true, Asset: asset.Spot, Authenticated: true, Channel: subscription.MyTradesChannel},
+	// Authenticated channels are currently being managed by the `generateAuthSubscriptions` method for the private connection
```
They will silently cease working, I have applied your suggestion thank you ser. 🖖

shazbert replied:
```console
[WARN]  | WEBSOCKET | 06/08/2025 10:58:52 | "Bybit" has an authenticated subscription "myOrders" in config which is not supported. Please remove
[WARN]  | WEBSOCKET | 06/08/2025 10:58:52 | "Bybit" has an authenticated subscription "myWallet" in config which is not supported. Please remove
[WARN]  | WEBSOCKET | 06/08/2025 10:58:52 | "Bybit" has an authenticated subscription "myTrades" in config which is not supported. Please remove
```

gloriousCode commented on exchanges/bybit/bybit_websocket.go:
```diff
@@ -63,94 +65,61 @@ var defaultSubscriptions = subscription.List{
 	subscription.MyOrdersChannel:  chanOrder,
-	subscription.MyTradesChannel:  chanExecution,
 	subscription.MyWalletChannel:  chanWallet,
+	subscription.MyTradesChannel:  chanExecution,
```
[ERROR] | WEBSOCKET | 07/08/2025 07:41:11 | Bybit Error subscribing subscribe with request ID 2 msg: Invalid topic :[execution.spot]

I've been getting this using both subscriptions from master and generated ones on this branch

gloriousCode reviewed (changes requested):
I think once the bug is addressed I'm happy with the work

gloriousCode commented on exchanges/bybit/bybit_websocket.go:
```diff
@@ -793,3 +674,171 @@ const subTplText = `
+
+	for _, configSub := range e.Config.Features.Subscriptions.Enabled() {
+		if configSub.Authenticated {
+			log.Warnf(log.WebsocketMgr, "%q has an authenticated subscription %q in config which is not supported. Please remove", e.Name, configSub.Channel)
```
```suggestion
			log.Warnf(log.WebsocketMgr, "%q has an authenticated subscription %q in config which is not supported. Please remove", e.Name, configSub.Channel)
			configSub.Enabled = false
```

gloriousCode commented on exchanges/bybit/bybit_websocket.go:
```diff
@@ -755,7 +636,7 @@ func channelName(s *subscription.Subscription) string {
+	case chanPositions, chanExecution, chanOrder, chanWallet:
 		return false
 	}
 	return true
```
Update `subTplText` to not process auth endpoints:
```
{{ with $name := channelName $.S }}
	{{- range $asset, $pairs := $.AssetPairs }}
		{{- if isSymbolChannel $name }}
			{{- range $p := $pairs }}
				{{- $name -}} .
				{{- if eq $name "orderbook" -}} {{- $.S.Levels -}} . {{- end }}
				{{- if eq $name "kline" -}} {{- intervalToString $.S.Interval -}} . {{- end }}
				{{- $p }}
				{{- $.PairSeparator }}
			{{- end }}
		{{- end }}
	{{- end }}
	{{- $.AssetSeparator }}
{{- end }}
```

shazbert replied:
https://github.com/thrasher-corp/gocryptotrader/pull/1670/commits/80a5edb11e8b8d4093df7c3d133def5dc2b45b60

gloriousCode reviewed (changes requested):
This should fix your issues

gloriousCode reviewed (approved):
i am pleased. tACK

thrasher- commented on exchanges/bybit/bybit_websocket.go:
```diff
@@ -793,3 +671,172 @@ const subTplText = `
+
+	for _, configSub := range e.Config.Features.Subscriptions.Enabled() {
+		if configSub.Authenticated {
+			log.Warnf(log.WebsocketMgr, "%q has an authenticated subscription %q in config which is not supported. Please remove", e.Name, configSub.Channel)
```
```suggestion
			log.Warnf(log.WebsocketMgr, "%s has an authenticated subscription %q in config which is not supported. Please remove.", e.Name, configSub.Channel)
```

Just a consistency nit since we generally use the exchange name without quotes

thrasher- commented on common/common.go:
```diff
@@ -669,15 +669,15 @@ func SortStrings[S ~[]E, E fmt.Stringer](x S) S {
 // Counter is a thread-safe counter.
 type Counter struct {
-	n int64 // privatised so you can't use counter as a value type
+	n atomic.Int64 // privatised so you can't use counter as a value type
```
A wise man once said:

"This realignment due to the requirement for assignment alignment brings no excitement, only chastisement that my review indictment brought defilement"

Howver this change rids the need for such things 😆

gbjk commented on exchanges/bybit/bybit_websocket.go:
```diff
@@ -238,71 +189,59 @@ func (by *Bybit) wsHandleData(_ context.Context, assetType asset.Item, respRaw [
+		if result.RequestID != "" {
+			return conn.RequireMatchWithData(result.RequestID, respRaw)
+		}
+
```
🦍 
```suggestion
```

gbjk commented on exchanges/bybit/bybit_websocket.go:
```diff
@@ -730,9 +661,6 @@ const subTplText = `
 			{{- end }}
-		{{- else }}
-			{{- $name }}
-			{{- if and (isCategorisedChannel $name) ($categoryName := getCategoryName $asset) -}} . {{- $categoryName -}} {{- end }}
```
🚧 I think this means we should be removing `isCategorisedChannel`.
I can't see it used anywhere. Sorry if I've missed it.

gbjk commented on exchanges/bybit/bybit_websocket.go:
```diff
@@ -891,10 +826,10 @@ func (by *Bybit) getPairFromCategory(category, symbol string) (currency.Pair, as
 		assets = append(assets, asset.Options)
 	default:
-		return currency.EMPTYPAIR, 0, fmt.Errorf("category %q not supported for incoming symbol %q", category, symbol)
+		return currency.EMPTYPAIR, 0, fmt.Errorf("incoming symbol %q %w: %q", symbol, errUnsupportedCategory, category)
```
🐒 This error feels back to front
```suggestion
		return currency.EMPTYPAIR, 0, fmt.Errorf("%w %q for symbol %q %w: %q", errUnsupportedCategory, category, symbol)
```

gbjk reviewed (changes requested):
Thanks for the updates. 

Just a few change requests and nitpicks.

shazbert commented on exchanges/bybit/bybit_test.go:
```diff
@@ -3051,31 +3073,43 @@ func TestPushDataPublic(t *testing.T) {
 
+	expErrFn := func(k string) error {
+		if k == "unhandled" {
+			return errUnhandledStreamData
+		}
+		return nil
+	}
+
 	for x := range keys {
-		err := e.wsHandleData(t.Context(), asset.Spot, []byte(pushDataMap[keys[x]]))
-		assert.NoError(t, err, "wsHandleData should not error")
+		assert.ErrorIsf(t, e.wsHandleData(nil, asset.Spot, []byte(pushDataMap[keys[x]])), expErrFn(keys[x]), "wsHandleData should not error for %s", keys[x])
```
much better

gbjk reviewed (approved):
Thanks for addressing those. I'm happy.

thrasher- reviewed (approved):
Changes ACK
